File: javascript/leetcode/690. Employee Importance.js
/**
 * Definition for Employee.
 * function Employee(id, importance, subordinates) {
 *     this.id = id;
 *     this.importance = importance;
 *     this.subordinates = subordinates;
 * }
 */

/**
 * @param {Employee[]} employees
 * @param {number} id
 * @return {number}
 */
var GetImportance = function (employees, id) {
  const seen = new Map();

  const dfs = (id) => {
    const emp = seen.get(id);

    let ans = emp.importance;

    for (let i of emp.subordinates) ans += dfs(i);

    return ans;
  };

  for (let i of employees) seen.set(i.id, i);

  return dfs(id);
};

File: c++/leetcode/690. Employee Importance.cpp
/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution
{
  unordered_map<int, Employee *> seen;

public:
  int dfs(int);
  int getImportance(vector<Employee *> employees, int id)
  {
    for (auto i : employees)
      seen.insert({i->id, i});

    return dfs(id);
  }
};

int Solution::dfs(int id)
{
  auto emp = seen[id];

  auto ans = emp->importance;

  for (auto i : emp->subordinates)
    ans += dfs(i);

  return ans;
}